[LoopVectorize] Add support for vectorisation of more early exit loops

david-arm opened the pull request:
This patch follows on from PR https://github.com/llvm/llvm-project/pull/107004 by adding support for vectorisation of a simple class of loops that typically involves searching for something, i.e.

```
  for (int i = 0; i < n; i++) {
    if (p[i] == val)
      return i;
  }
  return n;
```

or

```
  for (int i = 0; i < n; i++) {
    if (p1[i] != p2[i])
      return i;
  }
  return n;
```

In this initial commit we will only vectorise early exit loops legal if they
follow these criteria:

1. There are no stores in the loop.
2. The loop must have only one early uncountable exit like those shown in the
above example.
4. The early exit block dominates the latch block.
5. The latch block must have an exact exit count.
6. The loop must not contain reductions or recurrences.
7. We must be able to prove at compile-time that loops will not contain
faulting loads.

For point 7 once this patch lands I intend to follow up by supporting
some limited cases of faulting loops where we can version the loop based
on pointer alignment. For example, it turns out in the SPEC2017 benchmark
(xalancbmk) there is a std::find loop that we can vectorise provided we
add SCEV checks for the initial pointer being aligned to a multiple of
the VF. In practice, the pointer is regularly aligned to at least 32/64
bytes and since the VF is a power of 2, any vector loads <= 32/64 bytes
in size will always fault on the first lane, following the same behaviour
as the scalar loop. Given we already do such speculative versioning for
loops with unknown strides, alignment-based versioning doesn't seem to be
any worse at least for loops with only one load.

This patch makes use of the existing experimental_cttz_elems intrinsic
that's required in the vectorised early exit block to determine the first
lane that triggered the exit. This intrinsic has generic lowering support
so it's guaranteed to work for all targets.

Tests have been updated here:

Transforms/LoopVectorize/simple_early_exit.ll

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10260,7 +10562,11 @@ bool LoopVectorizePass::processLoop(Loop *L) {
   }
 
-  assert(!verifyFunction(*L->getHeader()->getParent(), &dbgs()));
+  //  assert(!verifyFunction(*L->getHeader()->getParent(), &dbgs()));
```
My apologies - I just realised this debug code is still present. I'll fix asap!

huntergr-arm commented on llvm/include/llvm/Support/GenericLoopInfo.h:
```diff
@@ -294,6 +294,10 @@ template <class BlockT, class LoopT> class LoopBase {
   BlockT *getUniqueExitBlock() const;
 
+  /// Return the exit block for the latch if one exists. This function assumes
+  /// the loop has a latch.
```
I think all loops have at least one latch (a branch back to the header of the loop). Perhaps 'This function assumes the loop has a single latch.'?

huntergr-arm commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -2423,6 +2571,13 @@ void LoopAccessInfo::analyzeLoop(AAResults *AA, LoopInfo *LI,
       if (I.mayWriteToMemory()) {
         auto *St = dyn_cast<StoreInst>(&I);
+        if (SpeculativeEarlyExitingBB) {
+          recordAnalysis("CantVectorizeInstruction", St)
```
What happens if St is null here?

huntergr-arm commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -2567,6 +2729,13 @@ void LoopAccessInfo::analyzeLoop(AAResults *AA, LoopInfo *LI,
   Accesses.buildDependenceSets();
 
+  if (SpeculativeEarlyExitingBB) {
+    assert(!Stores.size() && "Did not expect stores in an early exit loop!");
```
I think this effectively duplicates the previous assert?

huntergr-arm commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -8592,6 +8595,66 @@ ScalarEvolution::BackedgeTakenInfo::getExact(const Loop *L, ScalarEvolution *SE,
+  if (!Latch)
+    return SE->getCouldNotCompute();
+
+  if (!hasAnyInfo())
```
You can combine this with the above '!Latch' check.

arcbbb commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2831,6 +2840,18 @@ class VPRegionBlock : public VPBlockBase {
   /// VPRegionBlock.
   VPBlockBase *Exiting;
 
+  /// Hold the Early Exiting block of the SEME region, if one exists.
+  VPBlockBase *EarlyExit;
```
IIUC, for the SEME region, the latch exit VPBlock `middle.block` is placed as the successor of the region block, and the other exit VPBlocks other than the latch exit VPBlock are included in the region block, e.g. `vector.early.exit` pointed by `EarlyExit`.
And the early exiting block would be the predecessor of `EarlyExit`. 
Therefore the comment should be updated to `/// Hold the Early Exit block of the SEME region, if one exists`.

nikic commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -8592,6 +8595,63 @@ ScalarEvolution::BackedgeTakenInfo::getExact(const Loop *L, ScalarEvolution *SE,
+  // a later poison exit count should not propagate into the result. This are
+  // exactly the semantics provided by umin_seq.
+  return SE->getUMinFromMismatchedTypes(Ops, /* Sequential */ true);
+}
```
How does this "speculative" BECount differ from the SymbolicMax BECount?

david-arm replied:
Hmm, it does seem like it does the same thing, except there is no predicated version that accepts a vector of SCEVPredicate pointers, which is required for getPredicatedBackedgeTakenCount. I can try adding a predicated version of getSymbolicMax to see if that works.

david-arm replied:
The other major difference between getSpeculative and getSymbolicMax is the former requires the exact-not-taken count for the latch to be known, whereas the latter doesn't care. So I think in order to use something close to the existing getSymbolicMax interface I'll need to do two things:

1. Rewrite getSymbolicMax (or add an overloaded interface) so that it's a `const` interface (allowing it to be called from getPredicatedBackedgeTakenCount). Also, add a `SmallVector<const SCEVPredicate *, 4> *Predicates` argument.
2. Add code to getPredicatedBackedgeTakenCount to explicitly check we have a exact-not-taken count for the latch.

I'm happy to do this of course - just pointing out that getSymbolicMax isn't a drop-in replacement that's all. I'll try it out and see if I get the same behaviour as before.

david-arm replied:
I've tried posting a new commit that teaches getPredicatedBackedgeTakenCount to use a version of getSymbolicMax that accepts predicates, provided we have an exact count for the latch. Hopefully this makes better reuse of the code.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -3461,16 +3611,28 @@ void InnerLoopVectorizer::fixVectorizedLoop(VPTransformState &State,
     for (const auto &Entry : Legal->getInductionVars())
       fixupIVUsers(Entry.first, Entry.second,
                    getOrCreateVectorTripCount(VectorLoop->getLoopPreheader()),
-                   IVEndValues[Entry.first], LoopMiddleBlock,
-                   VectorLoop->getHeader(), Plan, State);
+                   IVEndValues[Entry.first], LoopMiddleBlock, Plan, State);
   }
 
   // Fix LCSSA phis not already fixed earlier. Extracts may need to be generated
   // in the exit block, so update the builder.
+  BasicBlock *OrigEarlyExitBB = Legal->hasSpeculativeEarlyExit()
+                                    ? Legal->getSpeculativeEarlyExitBlock()
+                                    : nullptr;
+
   State.Builder.SetInsertPoint(State.CFG.ExitBB,
                                State.CFG.ExitBB->getFirstNonPHIIt());
   for (const auto &KV : Plan.getLiveOuts())
-    KV.second->fixPhi(Plan, State);
+    if (OrigLoop->getUniqueExitBlock() ||
+        KV.second->getPhi()->getParent() != OrigEarlyExitBB)
+      KV.second->fixPhi(Plan, State);
```
Sorry I missed this before - I'll take a look!

david-arm replied:
Good spot! This code was just a hangover from an earlier implementation and it's no longer necessary to have the if guards - we can unconditionally handle all the live-outs.

sjoerdmeijer commented on llvm/test/Transforms/LoopVectorize/AArch64/simple_early_exit.ll:
```diff
@@ -0,0 +1,2544 @@
+; CHECK:       loop.end:
+; CHECK-NEXT:    [[START_0_LCSSA:%.*]] = phi i64 [ [[INDEX]], [[LAND_RHS]] ], [ 43, [[FOR_INC]] ], [ [[IND_EARLY_ESCAPE]], [[VECTOR_EARLY_EXIT]] ], [ 43, [[MIDDLE_BLOCK]] ]
+; CHECK-NEXT:    ret i64 [[START_0_LCSSA]]
+;
```
Hi @david-arm , we have already talked about this a little, but I was wondering about the profitability of vectorising these cases. This is an example of a loop with a very low trip count, and I cannot imagine that vectorisation of this particular case will be very beneficial. Do you think think we might need some heuristics and profitability checks? Do you have data and/or opinions on this?

david-arm replied:
That's a good question! Well, the program `xz` is one such example where we already vectorise an early-exit loop using the AArch64LoopIdiomTransform pass, where the loop is known to have low trip counts. That pass uses a predicated loop as the most optimal solution, but even an unpredicated vector loop with a scalar epilogue was still seen as an improvement on the scalar loop. So it's not entirely obvious to me that we shouldn't be vectorising loops with low trip counts.

I think ultimately we should let the vectoriser decide the best approach based on the cost model, and there is already code in the vectoriser that takes into account the cost of runtime checks and tries to determine the minimum trip count (if known at compile time) required to be profitable. Perhaps as part of this work I could add the cost of the cttz intrinsic in a similar way so that it can make a more informed choice?

david-arm replied:
I've added the cost of work required in the early exit block when assessing profitability of vectorising. This will help define the minimum profitable trip count required in order to enter the vector loop.

david-arm replied:
I will independently create a separate patch that improves the cost model for the cttz.elts intrinsic. For now I've disabled early-exit autovectorisation until we've got a better cost model, but that shouldn't hold up this patch.

sjoerdmeijer replied:
Hi, sorry for the late reply. I wanted to elaborate a little bit about why I brought up cost-modeling. It's kind of all in the name with "simple early exit": I was wondering if vectorisation would be beneficial if loops exit early. We now have two code paths: the early exit and the full vector body, which makes the perfomance modeling and characteristics let's say less uniform. That's why I wrote a couple of micro benchmark to test this, i.e. I tried to find at what iteration taking the early exit would give a performance uplift. For simple loops, I found you'll need at least a couple of hundred iterations for this vectorisation strategy to be beneficial. This is where my first observation of low trip counts came from: when we execute just a couple of iterations, the overhead of vectorisation is not worth it and actually results in (minor) regressions compared to scalar code. 

I hope I explained this a little bit better now. And by the way, there's probably a lot more going on than just looking at low trip count, but again this was just a first observation.

When the trip count isn't known at compile time, there's nothing we can do, I think. But when the trip count is a small constant, like some of the regression test cases, that looks like something we could handle.

I haven't looked at your latest changes related to this, just wanted to leave this comment first, but will take a look.

david-arm commented on llvm/include/llvm/Analysis/LoopAccessAnalysis.h:
```diff
@@ -696,6 +721,17 @@ class LoopAccessInfo {
+  bool LoopMayFault = false;
+
+  /// Keeps track of the early-exiting block, if present.
+  BasicBlock *SpeculativeEarlyExitingBB = nullptr;
```
In a sense though that is not far off the truth, because when vectorising the loop we are by definition reading ahead in memory which could potentially cause a fault where the scalar loop would not. However, the main reason I added the word 'Speculative' was to distinguish between early exits with exact exit-not-taken counts (which the vectoriser does support) and early exits that cannot be counted.

I'd prefer not to call it `EarlyExitingBB` to avoid any possible confusion, but I'm happy to take suggestions on alternative names that are better? Perhaps `UncountableEarlyExitingBB`?

david-arm commented on llvm/include/llvm/Analysis/LoopAccessAnalysis.h:
```diff
@@ -696,6 +721,17 @@ class LoopAccessInfo {
+
+  /// Keeps track of all the early exits with known or countable exit-not-taken
+  /// counts.
+  SmallVector<BasicBlock *, 4> CountableEarlyExitingBlocks;
```
That makes sense!

sjoerdmeijer commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9928,15 +9928,74 @@ static void checkMixedPrecision(Loop *L, OptimizationRemarkEmitter *ORE) {
+      }
+    }
+
+    for (User *U : OrigPhi->users()) {
```
Nit: this could be lamba, to share code with the previous for-loop that is almost the same?

sjoerdmeijer commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9928,15 +9928,74 @@ static void checkMixedPrecision(Loop *L, OptimizationRemarkEmitter *ORE) {
+      auto *UI = cast<Instruction>(U);
+      if (!L->contains(UI)) {
+        PHINode *PHI = dyn_cast<PHINode>(UI);
+        assert(PHI && "Expected LCSSA form");
```
Nit: checking LCSSA form could be hoisted and checked earlier and just once?

david-arm replied:
I can't hoist this assert out, since it's based upon the User `U` which varies in each loop iteration.

sjoerdmeijer commented:
I was wondering if it would be good to add some AArch64 codegen tests too so that we can look at some codegen?

david-arm commented:
> I was wondering if it would be good to add some AArch64 codegen tests too so that we can look at some codegen?

If you're referring to the codegen coming out of clang after vectorising the loop, I don't think we typically have tests like that in test/Transform/LoopVectorize. They are normally IR/opt based tests. Are you referring specifically to the codegen from the cttz.elts intrinsic? If so, we already have tests for them - see CodeGen/AArch64/intrinsic-cttz-elts-sve.ll.

sjoerdmeijer commented:
> > I was wondering if it would be good to add some AArch64 codegen tests too so that we can look at some codegen?
> 
> If you're referring to the codegen coming out of clang after vectorising the loop, I don't think we typically have tests like that in test/Transform/LoopVectorize. They are normally IR/opt based tests. Are you referring specifically to the codegen from the cttz.elts intrinsic? If so, we already have tests for them - see CodeGen/AArch64/intrinsic-cttz-elts-sve.ll, for example.

Yes, I appreciate we test all things individually, but I was just thinking that it is a bit of shame we can't look at some codegen for a loop for all of this work. For example, take the resulting IR of some of the tests in test/Transform/LoopVectorize/AArch64, and create llc tests. Not sure if there's precedent for that, I guess not.

fhahn commented:
> Yes, I appreciate we test all things individually, but I was just thinking that it is a bit of shame we can't look at some codegen for a loop for all of this work. For example, take the resulting IR of some of the tests in test/Transform/LoopVectorize/AArch64, and create llc tests. Not sure if there's precedent for that, I guess not.

It would probably make sense to have some micro-benchmarks for some loops with varying trip counts (both statically known and unknown) to cover the end-to-end flow and allow for easy evaluation. Sharing the generated assembly end-to-end for some of those might help, as @sjoerdmeijer suggested?

(I don't think we should add end-to-end tests to llvm-project/llvm/tests/ directly that run the vectorizer (and possibly other passes) all the way down to assembly)

sjoerdmeijer reviewed (approved):
This patch basically contains two parts: the LAA/SCEV and the vectorisation part.

I have only looked at the vectorisation part and that looks good to me:
- thanks for taking the cost-model remarks into account, the added logic seems like a good first step,
- the option to vectorise loops with early breaks is off by default. This allows us to experiment more with this, possibly refine the cost-model, without creating a lot of turbulence. 
- It's a shame we can't look at final codegen for these sort of patches, but that is not a problem of this patch. I like the idea of some microbenchmarks for this, but given that this is off by default I don't think that this needs to hold up this patch.

So, LGTM, but I haven't looked at the LAA part, perhaps @nikic or @nikolaypanchenko can sign off on that part.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -181,6 +181,16 @@ static cl::opt<bool> EnableEpilogueVectorization(
+    cl::desc("Enable vectorization of early exit loops."));
+
+static cl::opt<bool> AssumeNoMemFault(
+    "vectorizer-no-mem-fault", cl::init(false), cl::Hidden,
```
Is the intention here to use this for writing unit tests or something else? for unit tests, here are multiple ways to tell LLVM that memory is dereferenceable for a range (with a constant number of elements), which should cover most cases when writing tests?

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/simple_early_exit.ll:
```diff
@@ -0,0 +1,2660 @@
+;
+entry:
+  %p1 = alloca [1024 x i8]
+  %p2 = alloca [1024 x i8]
```
At the moment, the tests read from uninitialized memory, which means automatic verification won't work. It would probably also make them more robust to avoid reading from uninitialized memory (e.g. by passing the pointers to a call first)

david-arm replied:
Yeah sure I'll try to come up with a different title. I am aware the vectoriser already supports some simpler early exit cases, which I've made an effort to address in the patch in terms of functionality and terminology. In fact, this patch permits multiple early exits in a loop, provided only one of them has an unknown exit-not-taken count. In such cases a scalar epilogue is still required. Essentially we have to distinguish between loops with countable early exits and non-countable ones.

david-arm replied:
Done!

fhahn reviewed (commented):
LV already support vectorizing early-exit loops for arguably even simpler cases, i.e. the exit counts for all exits can be computed (in that case, the scalar epilogue is run at least once to fix) https://llvm.godbolt.org/z/35v3nzY7K. So it would probably be worth clarifying that in the commit title.


I added some initial comments/questions, but as this patch touches a lot of different things I wasn't yet able to complete a full look through. Hope to make it through all of it by early/mid next week.

david-arm commented:
I just did a simple rebase and fixed some failing tests by teaching collectLoopUniforms about early exits.

david-arm commented:
@sjoerdmeijer @fhahn Here is an example of what happens with this patch when I build the following C code with `-O3 -target aarch64-linux -mcpu=neoverse-v1 -S -mllvm -enable-early-exit-vectorization`:

```
int foo(unsigned char *vec, unsigned char val) {
  unsigned char local_vec[128];

  for (int i = 0; i < 128; i++) {
    local_vec[i] = vec[i] + i;
  }

  // ACTUAL EARLY EXIT LOOP!
  unsigned char *p = &local_vec[0];
  for (int i = 0; i < 128; i++) {
    if (p[i] == val)
      return i;
  }
  return -1;
}
```

The code looks a bit contrived because this patch will only vectorise if it can prove the loads in the early-exit loop will no fault. Here is the assembly for the vector portion of the loop:

```
.LBB0_16:
        rdvl    x10, #15
        mov     x9, xzr    
        and     x0, x10, #0x80
        mov     z0.b, w1
        ptrue   p0.b
        mov     x10, sp
        .p2align        5, , 16
.LBB0_17:
        ld1b    { z1.b }, p0/z, [x10, x9]
        cmpeq   p1.b, p0/z, z1.b, z0.b
        b.ne    .LBB0_21
        add     x9, x9, x8
        cmp     x0, x9
        b.ne    .LBB0_17
        cbz     x0, .LBB0_12
        mov     w0, #-1
        add     sp, sp, #128
        ret
.LBB0_21:
        brkb    p0.b, p0/z, p1.b
        incp    x9, p0.b
        mov     x0, x9
.LBB0_22:
        add     sp, sp, #128
        ret
```

Thanks!

david-arm commented:
I just realised this patch is missing a test for a reverse loop, which also works perfectly fine and generates the correct code. I'll put up a new patch today.

fhahn commented:
Starting to think about the code for the dependence/memory analysis, I was wondering if this could be generalized a bit more separately. AFAICT the dependence analysis and runtime check generation part of LAA does not really care if there are early exits, as  long as we can compute the symbolic max, which should be safe to use instead of the exact BTC: https://github.com/llvm/llvm-project/pull/93499

With that, it seems like `isAnalyzableEarlyExitLoop` and `mayFault` would be better placed outside LAA?

david-arm commented:
> Starting to think about the code for the dependence/memory analysis, I was wondering if this could be generalized a bit more separately. AFAICT the dependence analysis and runtime check generation part of LAA does not really care if there are early exits, as long as we can compute the symbolic max, which should be safe to use instead of the exact BTC: #93499
> 
> With that, it seems like `isAnalyzableEarlyExitLoop` and `mayFault` would be better placed outside LAA?

I'm on holiday at the moment I'm afraid, but I'm happy to review your comments and look at any patches you have when I get back on Wednesday next week.

david-arm commented:
> Starting to think about the code for the dependence/memory analysis, I was wondering if this could be generalized a bit more separately. AFAICT the dependence analysis and runtime check generation part of LAA does not really care if there are early exits, as long as we can compute the symbolic max, which should be safe to use instead of the exact BTC: #93499
> 
> With that, it seems like `isAnalyzableEarlyExitLoop` and `mayFault` would be better placed outside LAA?

I agree with trying to break this patch down and making LoopAccessAnalysis as generic and untied to the vectoriser as possible, but I will definitely need to do work in LoopAccessAnalysis to provide the right information for users of it. For example, LoopAccessAnalysis does all the analysis of loads and stores in the loop, not LoopVectorizationLegality where we now test for uncomputable backedge counts. We still need to determine whether or not the loop has an uncountable exit and, if so, whether there are any memory accesses after the exit, which is something more suited to LoopAccessAnalysis. If we move all the analysis into LoopVectorizationLegality then we basically have to duplicate the same work of walking through all the instructions in all the blocks that already happens in canAnalyzeLoop. In this new world I will still need to calculate the earliest uncountable exiting block in LoopAccessAnalysis and track post-exit memory accesses so that in LoopVectorizationLegality we can query the data to make an informed choice. I'll look into doing that, but the main point is that we can't devolve all work into LoopVectorizationLegality.

I may be able to move mayFault, but it does depend upon structures built up by LoopAccessAnalysis and I personally thought that this function might be useful outside of the vectoriser so it wasn't obvious where to move it to.

fhahn commented:
> I may be able to move mayFault, but it does depend upon structures built up by LoopAccessAnalysis and I personally thought that this function might be useful outside of the vectoriser so it wasn't obvious where to move it to.

One place that might be worth looking at is https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp#L1343. `isDereferenceableAndAlignedInLoop` already has some logic to check if an access is dereferenceable across all loop iterations, it may be able to replace parts of the current `mayFault` implementation?

david-arm commented:
> > I may be able to move mayFault, but it does depend upon structures built up by LoopAccessAnalysis and I personally thought that this function might be useful outside of the vectoriser so it wasn't obvious where to move it to.
> 
> One place that might be worth looking at is https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp#L1343. `isDereferenceableAndAlignedInLoop` already has some logic to check if an access is dereferenceable across all loop iterations, it may be able to replace parts of the current `mayFault` implementation?

OK thanks for the info. At the moment I've moved it into the LoopInfo class because it's specific to loops and I thought it was potentially useful outside of the vectoriser. You still have to record all of the underlying objects that were collected by LoopAccessAnalysis and pass those in. It seems the only real benefit for moving this code outside of LoopAccessAnalysis (which already has all the information we need) is to make it more generally available to all, rather than tying it to the vectoriser. What do you think?

david-arm commented:
I have rebased on top of HEAD of LLVM, which contains some of Florian's recent changes to LoopAccessAnalysis to generalise all loops to use the symbolic maximum. This required a significant rewrite of the patch, and rebasing/merging to keep all of the commit history was practically impossible. Instead I've simply forced an update with a single new commit.

david-arm commented:
Hmm, some changes in LoopAccessAnalysis.cpp seem to be missing. I'm taking a look.

fhahn commented on llvm/include/llvm/Analysis/LoopAccessAnalysis.h:
```diff
@@ -653,7 +659,24 @@ class LoopAccessInfo {
 
   unsigned getNumStores() const { return NumStores; }
-  unsigned getNumLoads() const { return NumLoads;}
+  unsigned getNumLoads() const { return NumLoads; }
```
unused?

fhahn commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -2353,6 +2349,36 @@ void MemoryDepChecker::Dependence::print(
+  TheLoop->getExitingBlocks(ExitingBlocks);
+
+  SmallVector<BasicBlock *, 4> CountableExitingBBs;
+  PSE->getSE()->getCountableExitingBlocks(TheLoop, &CountableExitingBBs);
```
If getCountableExitingBlocks is only used here, could it be moved out of LAI?

fhahn commented on llvm/include/llvm/Analysis/LoopAccessAnalysis.h:
```diff
@@ -653,7 +659,24 @@ class LoopAccessInfo {
+  }
+
+  /// Returns all the exiting blocks with an uncountable exit.
+  const SmallVector<BasicBlock *, 4> &getUncountableExitingBlocks() const {
```
Only used in LoopVectorize/LoopVectorizationLegality, could this be moved there?

fhahn replied:
IMO LAA is already quite complicated and should focus on checking dependences and collecting infos for runtime checks and it would be better to avoid introducing complexity which may be used by other passes in the unknown future. Keeping LAA focused hopefully makes it easier to extend its capabilities w.r.t. to generating runtime checks and dependence analysis  in the future

As the functionality is independent of LAA, it could live in a separate utility if decoupling of LV is desired.

fhahn commented on llvm/include/llvm/Analysis/LoopAccessAnalysis.h:
```diff
@@ -653,7 +659,24 @@ class LoopAccessInfo {
+  }
+
+  /// Returns all the exit blocks from uncountable exiting blocks.
+  SmallVector<BasicBlock *, 4> getUncountableExitBlocks() const {
```
Only used in LoopVectorize/LoopVectorizationLegality, could this be moved there?

fhahn commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -2514,14 +2545,20 @@ bool LoopAccessInfo::analyzeLoop(AAResults *AA, LoopInfo *LI,
+    // Don't return yet if there are uncountable exits because we also need to
+    // record all of the underlying objects for memory fault analysis.
+    if (!UncountableExitingBlocks.size())
+      return true;
```
Relying on analyzeLoop to collect all underlying objects is fragile (it is easy to miss this invariant when a new exit is added) and adds additional complexity to LAA which isn't needed for its job. It may also have to iterate over the whole loop for users that don't support loops with uncountable exits.

We already iterate over the loop multiple times in LoopVectorizationLegality to collect information needed to decide whether to vectorize, so IMO it would make sense to move the logic there, instead adding complexity to LAA which isn't needed for dependence analysis/runtime check generation.

As mentioned earlier, we also already collect pointers we know won't fault across all iterations of the loop. Would be good to reuse/extend the existing logic

fhahn commented on llvm/lib/Analysis/VectorUtils.cpp:
```diff
@@ -1489,4 +1489,82 @@ void InterleaveGroup<Instruction>::addMetadata(Instruction *NewInst) const {
+    //   getelementptr [32 x i32], ptr %arr, i64 0, i64 %ind
+    Value *GEPInd;
+    Type *GEPElemType;
+    if (GEP->getNumIndices() == 2) {
```
I'm not sure if this is correct; shouldn't we use the accessed type here? (likely needs a test)

fhahn commented on llvm/lib/Analysis/VectorUtils.cpp:
```diff
@@ -1489,4 +1489,82 @@ void InterleaveGroup<Instruction>::addMetadata(Instruction *NewInst) const {
+        return true;
+      GEPInd = GEP->getOperand(2);
+      GEPElemType = ArrayTy->getElementType();
+    } else if (GEP->getNumIndices() == 1) {
```
Why are those restrictions needed when we later use the SCEV of the GEP? If possible, would be good to use isDereferenceableAndAlignedInLoop instead of duplicating its logic.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1054,9 +1060,11 @@ bool LoopVectorizationLegality::canVectorizeInstrs() {
-  LAI = &LAIs.getInfo(*TheLoop);
+bool LoopVectorizationLegality::canVectorizeMemory(bool IsEarlyExitLoop) {
+  if (!LAI)
+    LAI = &LAIs.getInfo(*TheLoop);
```
Why is this needed?

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1054,9 +1060,11 @@ bool LoopVectorizationLegality::canVectorizeInstrs() {
+  if (!LAI)
+    LAI = &LAIs.getInfo(*TheLoop);
   const OptimizationRemarkAnalysis *LAR = LAI->getReport();
+
```
unrelated change

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1439,6 +1498,74 @@ bool LoopVectorizationLegality::canVectorizeLoopNestCFG(
+  }
+
+  if (!LAI)
+    LAI = &LAIs.getInfo(*TheLoop);
```
why is this need?

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -979,8 +987,10 @@ Value *getRuntimeVF(IRBuilderBase &B, Type *Ty, ElementCount VF) {
-  const SCEV *BackedgeTakenCount = PSE.getBackedgeTakenCount();
+                                Loop *OrigLoop, bool IsSpeculative) {
+  const SCEV *BackedgeTakenCount = IsSpeculative
+                                       ? PSE.getSymbolicMaxBackedgeTakenCount()
```
If there's a exact BTC then symbolic Max BTC == exact BTC, so the distinction here shouldn't be needed?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -677,9 +678,14 @@ class VPBlockBase {
 
+  /// Is this a live-out value specifically for an early exit from the vector
+  /// loop? If so, it needs handling specially.
+  bool EarlyExit;
```
Why is this needed? Would it be sufficient to record the predecessor as done in #94760? This would be more general in handling live-outs in different VPBBs

david-arm replied:
Possibly? #94760 is new to me so I'll have to take a look.

david-arm replied:
It looks like that #94760 could take a while to land. I'd prefer not to make this PR dependent on that, unless you are saying that I take a small part of #94760?

david-arm reviewed (commented):
Hi @fhahn, thanks for taking a look. I've left some replies and I'll take a look at refactoring LoopVectorizationLegality to pull some of the logic out into a common place in lib/Analysis.

huntergr-arm reviewed (commented):
Overall, I think I'd like to see the LoopVectorize changes split out, and if possible some analysis tests written.

Maybe if `loopMayFault` was in LAA (if rewritten to analyze accesses), or in LoopInfo, you could write some tests to show loops with and without faults.

fhahn reviewed (commented):
> I have put this in VectorUtils.h for now, because I do see value in using this function outside of LoopVectorize. If anyone can think of a more suitable place I am happy to move it!

Is there a reason why the custom logic to check for faults is needed and  the exiting isDereferenceableAndAlignedInLoop cannot be used?

kiranchandramohan reviewed (commented):
A few trivial Nit comments.

david-arm commented:
> isDereferenceableAndAlignedInLoop

Currently `isDereferenceableAndAlignedInLoop` does not support reverse loops, which is a regression compared to my current patch. I can see how much work is required to make it work for negative step recurrences.

> isDe

david-arm commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -8604,6 +8604,24 @@ ScalarEvolution::BackedgeTakenInfo::getExact(const Loop *L, ScalarEvolution *SE,
+void ScalarEvolution::BackedgeTakenInfo::getCountableExitingBlocks(
+    const Loop *L, ScalarEvolution *SE,
+    SmallVector<BasicBlock *, 4> *Blocks) const {
+  // All exiting blocks we have collected must dominate the only backedge.
```
This was just an errant comment left over from a previous version. I've removed it now.

david-arm commented:
Hi @fhahn @huntergr-arm, I've pushed up a new patch which required fixing conflicts during rebase (again). Although I see it conflicts yet again - hard to keep standing still! :(

This patch now effectively contains #96752 and I will also create a new PR for the standalone addition of `isDereferenceableReadOnlyLoop`. Once these two separate PRs land it will make this patch a bit smaller.

Almost all of the LoopAccessAnalysis changes have gone, with the logic for determining whether a loop will fault moved to LoopVectorizationLegality. Similarly, the code to track exiting blocks has moved there too.

david-arm commented:
I've still not addressed all the most recent round of comments on this patch because I'm waiting for https://github.com/llvm/llvm-project/pull/96752 to land first. However, I've rebased the patch and included https://github.com/llvm/llvm-project/pull/96752 anyway to keep it ticking over. Every rebase usually takes a while due to changes in LoopVectorize/VPlan. :)

From tomorrow I'll be away for 3 weeks so apologies if I'm silent during that time, but feel free to leave more review comments and I'll take a look when I'm back!

huntergr-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1439,6 +1496,103 @@ bool LoopVectorizationLegality::canVectorizeLoopNestCFG(
+    }
+  }
+
+  CountableExitingBlocks = std::move(CountableExitingBBs);
```
No need to create temporaries and move the data, you can just add directly to the member vectors in LVL.

huntergr-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1439,6 +1496,103 @@ bool LoopVectorizationLegality::canVectorizeLoopNestCFG(
+  TheLoop->getExitingBlocks(ExitingBlocks);
+
+  SmallVector<BasicBlock *, 4> CountableExitingBBs;
+  PSE.getSE()->getCountableExitingBlocks(TheLoop, &CountableExitingBBs);
```
I don't think we need to add this method to ScalarEvolution; I tried the following and all tests still passed:

```
  for (BasicBlock *BB1 : ExitingBlocks)
    if (isa<SCEVCouldNotCompute>(PSE.getSE()->getExitCount(TheLoop, BB1)))
      UncountableExitingBlocks.push_back(BB1);
    else
      CountableExitingBlocks.push_back(BB1);
```

If there is a difference in semantics (maybe about the predicates?) then could you please add a test to demonstrate.

huntergr-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1439,6 +1496,103 @@ bool LoopVectorizationLegality::canVectorizeLoopNestCFG(
+  LLVM_DEBUG(
+      dbgs()
+      << "LV: Found an early exit. Retrying with speculative exit count.\n");
+  if (!PSE.getSE()->hasExactPredicatedBackedgeTakenCount(TheLoop, LatchBB)) {
```
I don't think this ScalarEvolution method needs to be added either.
```suggestion
  if (isa<SCEVCouldNotCompute>(PSE.getSE()->getExitCount(TheLoop, LatchBB)))
```

huntergr-arm reviewed (commented):
Starting to review the core representation in VPlan now; in the meantime, I think we can cut down on overall diff size by removing the ScalarEvolution changes.

huntergr-arm commented on llvm/include/llvm/Transforms/Vectorize/LoopVectorizationLegality.h:
```diff
@@ -404,6 +424,22 @@ class LoopVectorizationLegality {
 
+  /// Returns all exiting blocks with a countable exit, i.e. the
+  /// exit-not-taken count is known exactly at compile time.
+  const SmallVector<BasicBlock *, 4> &getCountableExitingBlocks() const {
```
I think it would be better to use `SmallVectorImpl<BasicBlock *>` instead of tying the result type to the default parameter.

huntergr-arm reviewed (commented):
I think it should be possible to split out the VPlan changes and test via llvm/unittests/Transforms/Vectorize/VPlanTest.cpp (and the other files in that directory).

I've experimented with this implementation a bit, and it doesn't quite model a single-entry-multiple-exit region. The vector early exit block seems like it's part of the vector region in the plan, when it should be external to it as it's not part of the loop. Separating out the vplan changes should let us see this and figure out how to represent the CFG properly.

david-arm commented:
Please note this patch is becoming out of date, since I've been splitting off various parts of this patch and landing them separately. Once enough of those patches have landed I will rebase this and hopefully the diff will be smaller and more concentrated on the vectoriser/vplan code!

david-arm commented:
* Tidied up requiresScalarEpilogue and fixupIVUsers
* Added the original early exit block to State.CFG as a more appropriate place than the VPRegionBlock.
* Cleaned up the createInitialVPlan interface so now we only pass in a boolean indicating whether we have an early exit or not.
* Updated VPLiveOut::fixPhi to use the new incoming block parameter instead of trying to reverse engineer the CFG.

david-arm committed:
[LoopVectorize] Add support for vectorisation of more early exit loops

This patch follows on from PR #107004 by adding support for
vectorisation of a simple class of loops that typically involves
searching for something, i.e.

  for (int i = 0; i < n; i++) {
    if (p[i] == val)
      return i;
  }
  return n;

or

  for (int i = 0; i < n; i++) {
    if (p1[i] != p2[i])
      return i;
  }
  return n;

In this initial commit we will only vectorise early exit loops legal
if they follow these criteria:

1. There are no stores in the loop.
2. The loop must have only one early uncountable exit like those shown
in the above example.
3. The early exit block dominates the latch block.
4. The latch block must have an exact exit count.
6. The loop must not contain reductions or recurrences.
7. We must be able to prove at compile-time that loops will not contain
faulting loads.

For point 7 once this patch lands I intend to follow up by supporting
some limited cases of faulting loops where we can version the loop based
on pointer alignment. For example, it turns out in the SPEC2017 benchmark
(xalancbmk) there is a std::find loop that we can vectorise provided we
add SCEV checks for the initial pointer being aligned to a multiple of
the VF. In practice, the pointer is regularly aligned to at least 32/64
bytes and since the VF is a power of 2, any vector loads <= 32/64 bytes
in size will always fault on the first lane, following the same behaviour
as the scalar loop. Given we already do such speculative versioning for
loops with unknown strides, alignment-based versioning doesn't seem to be
any worse at least for loops with only one load.

This patch makes use of the existing experimental_cttz_elems intrinsic
that's required in the vectorised early exit block to determine the first
lane that triggered the exit. This intrinsic has generic lowering support
so it's guaranteed to work for all targets.

Tests have been updated here:

Transforms/LoopVectorize/simple_early_exit.ll

david-arm commented:
Abandoned as most of the work has now landed, and is being enabled by default with https://github.com/llvm/llvm-project/pull/133099
